subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+			*cycle_cnt = 0;
+		} else if (rc) {
+			/* bad flash read */
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+/* zms_close_ate_valid validates an sector close ate: a valid sector close ate:
+ * - valid ate
+ * - len = 0 and id = ZMS_HEAD_ID
+ * - offset points to location at ate multiple from sector size
+ * return 1 if valid, 0 otherwise
+ */
+static int zms_close_ate_valid(struct zms_fs *fs, const struct zms_ate *entry)
+{
+	size_t ate_size;
+
+	if ((!zms_ate_valid_different_sector(fs, entry, entry->cycle_cnt)) || (entry->len != 0U) ||
+	    (entry->id != ZMS_HEAD_ID)) {
+		return 0;
+	}
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+	if ((fs->sector_size - entry->offset) % ate_size) {
+		return 0;
+	}
+
+	return 1;
+}
+
+/* zms_empty_ate_valid validates an sector empty ate: a valid sector empty ate:
+ * - valid ate
+ * - len = 0xffff and id = 0xffffffff
+ * return 1 if valid, 0 otherwise
+ */
+static int zms_empty_ate_valid(struct zms_fs *fs, const struct zms_ate *entry)
+{
+	if ((!zms_ate_valid_different_sector(fs, entry, entry->cycle_cnt)) ||
+	    (entry->len != 0xffff) || (entry->id != ZMS_HEAD_ID)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+/* store an entry in flash */
+static int zms_flash_write_entry(struct zms_fs *fs, uint32_t id, const void *data, size_t len)
+{
+	int rc;
+	struct zms_ate entry;
+
+	entry.id = id;
+	entry.len = (uint16_t)len;
+	entry.cycle_cnt = fs->sector_cycle;
+
+	if (len > 8) {
+		/* only compute CRC if len is greater than 8 bytes */
+		if (IS_ENABLED(CONFIG_ZMS_DATA_CRC)) {
+			entry.data_crc = crc32_ieee(data, len);
+		}
+		entry.offset = (uint32_t)(fs->data_wra & ADDR_OFFS_MASK);
+	} else if ((len > 0) && (len <= 8)) {
+		/* Copy data into entry for small data ( < 8B) */
+		uint8_t *data8 = (uint8_t *)data;
+
+		memcpy(&entry.data, data8, len);
+	}
+
+	zms_ate_crc8_update(&entry);
+
+	if (len > 8) {
+		rc = zms_flash_data_wrt(fs, data, len);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	rc = zms_flash_ate_wrt(fs, &entry);
+	if (rc) {
+		return rc;
+	}
+
+	return 0;
+}
+
+/* end of flash routines */
+
+/* Search for the last valid ATE written in a sector and also update data write address
+ */
+static int zms_recover_last_ate(struct zms_fs *fs, uint64_t *addr, uint64_t *data_wra)
+{
+	uint64_t data_end_addr, ate_end_addr;
+	struct zms_ate end_ate;
+	size_t ate_size;
+	int rc;
+
+	LOG_DBG("Recovering last ate from sector %llu", SECTOR_NUM(*addr));
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	/* skip close and empty ATE */
+	*addr -= 2 * ate_size;
+
+	ate_end_addr = *addr;
+	data_end_addr = *addr & ADDR_SECT_MASK;
+	while (ate_end_addr > data_end_addr) {
+		rc = zms_flash_ate_rd(fs, ate_end_addr, &end_ate);
+		if (rc) {
+			return rc;
+		}
+		if (zms_ate_valid(fs, &end_ate)) {
+			/* found a valid ate, update data_end_addr and *addr */
+			data_end_addr &= ADDR_SECT_MASK;
+			if (end_ate.len > 8) {
+				data_end_addr += end_ate.offset + end_ate.len;
+				*data_wra = data_end_addr;
```

Comment:
You should align this up to the write block size before assigning to `fs->data_wra` like NVS does. `fs->data_wra` must always be aligned to the write block size.

Reply:
data offset and length are already aligned before writing them to ATE.
This is just the recovery function.
Do you think it is worth adding an alignment here as well ?

Reply:
The thing is the length in the ATE is not aligned, right? It cannot be as the user must be able to restore the original data length.